RavenDB-22076 Vector search

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-22076

### Additional description

...Include details of the change made in this Pull Request or additional notes for the solution. Anything that can be useful for reviewers of this PR...

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [ ] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [ ] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [ ] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed 

### Testing by Contributor

- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [ ] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [ ] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [ ] No UI work is needed

redknightlois committed:
RavenDB-22261: Implementation of Portable Exceptions to simplify semantics.

ayende committed:
RavenDB-22076 Starting to add vector search to Corax

Currently just exploring what we can do

* Added Vector(byte[]) to accept PQ(1bit) vectors and store them in the index
* Added vector.search() that can accept a similar PQ(1bit) vector and execute Exact Nearest Neighbor search

ayende committed:
RavenDB-22076
- Implementing rough start of auto indexing vectors
- Optimize reading a lot of vectors from disk and cached the sorting results
- Auto indexes with vectors will default to using Corax
- Proper errors on Lucene engine when using vector search
- Handle vector.nearest() vs. vector.search()
- Allow to reuse a CompactKey to avoid allocations costs

ayende committed:
RavenDB-22076 Fixing rebase - avoiding NRE

Lwiel committed:
RavenDB-22076 Vector search client API

ayende committed:
RavenDB-22076
- Started to add HNSW implementation for Approximate Nearest Neighbor vector search
- Added content comparer for ByteString and allow to realloc from an empty ByteString
- Support reading & writing reversed variable size ints
- Added an easy way to call GenericHash from Sodium
- Removing unused methods in CompactTree and the entire unused ContainerState

maciejaszyk committed:
RavenDB-22076 Introduce:
- support for configuring embeddings/quantizaton in AutoIndexes
- support for configuring Vector field for static indexes
- Roslyn Rewriter for supporting creation of dynamic vector fields

maciejaszyk committed:
RavenDB-22076 Fixing code after rebase

maciejaszyk committed:
RavenDB-22076
- Unified vector creation via `CreateVector` method.
- Added support for sbytes and bytes embeddings in indexing and querying
- Unified vector method in query builder
- Created Vector configuration factory for setting up vector fields for static indexes
- Avoided allocation by passing `Memory<byte>` into `ExactSearchMatch` instead of copying buffer
- Fixed index definition writer
- Temporary added support for printing stored arrays of embeddings

maciejaszyk committed:
RavenDB-22076 Rebase fix

Lwiel committed:
RavenDB-22076 Added WithField method

maciejaszyk committed:
RavenDB-22076
- Unified `EmbeddingType` enum between Client API and Server API.
- Expanded embedding type in query building to support all numeric values.

maciejaszyk committed:
RavenDB-22076
- Unified VectorSearch method names between Server and Client.
- constant for the minimum similarity threshold.

Lwiel committed:
RavenDB-22076 Added IndexingStrategy

Lwiel committed:
RavenDB-22076 Removed TargetQuantization from WithField

maciejaszyk committed:
RavenDB-22076 Introducing support for indexing vectors in JavaScript indexes.

Lwiel committed:
RavenDB-22076 Adjusted exceptions, changed ExactVectorSearchMatch.SimilarityI8

maciejaszyk committed:
RavenDB-22076 Added Vector field for indexing in Raven.Studio.

Lwiel committed:
RavenDB-22076 Fixed AutoIndex tests

maciejaszyk committed:
RavenDB-22076 Performance: Encode memory to Base64 when using numeric arrays in the client.

maciejaszyk committed:
RavenDB-22076
- Renamed `Float32` to `Single` in `EmbeddingType`
- Fixed aliasing in static index field for querying (vector method)
- Separated CreateVector for the server-side embedding generator
- Added an exception when the user tries to define quantization while using `ByField` overload in VectorSearch

ppekrol commented on src/Raven.Client/Constants.cs:
```diff
@@ -577,5 +581,23 @@ private SupportedFeatures()
             }
         }
+        
+        public static class VectorSearch
```
internal

ppekrol commented on src/Raven.Client/Documents/Indexes/Vector/EmbeddingType.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace Raven.Client.Documents.Indexes.Vector;
+
+public enum EmbeddingType
```
VectorEmbeddingType?

ppekrol commented on src/Raven.Client/Documents/Indexes/Vector/VectorOptions.cs:
```diff
@@ -0,0 +1,112 @@
+        DestinationEmbeddingType = options.DestinationEmbeddingType;
+    }
+    
+    public static readonly VectorOptions Default = new()
```
internal?

ppekrol commented on src/Raven.Client/Documents/Indexes/Vector/VectorOptions.cs:
```diff
@@ -0,0 +1,112 @@
+        DestinationEmbeddingType = EmbeddingType.Single
+    };
+    
+    public static readonly VectorOptions DefaultText = new()
```
internal?

ppekrol commented on src/Raven.Client/Documents/Indexes/Vector/VectorOptionsFactory.cs:
```diff
@@ -0,0 +1,60 @@
+{
+    internal VectorOptions VectorOptions = options;
+
+    public VectorOptionsFactory Default => new VectorOptionsFactory(VectorOptions.Default);
```
internal?

ppekrol commented on src/Raven.Client/Documents/Indexes/Vector/VectorOptionsFactory.cs:
```diff
@@ -0,0 +1,60 @@
+
+    public VectorOptionsFactory Default => new VectorOptionsFactory(VectorOptions.Default);
+    
+    public VectorOptionsFactory() : this(new VectorOptions())
```
internal?

ppekrol commented on src/Raven.Client/Documents/Indexes/Vector/VectorOptionsFactory.cs:
```diff
@@ -0,0 +1,60 @@
+﻿namespace Raven.Client.Documents.Indexes.Vector;
+
+public sealed class VectorOptionsFactory(VectorOptions options)
```
ctor should be internal?

maciejaszyk replied:
yes

ppekrol commented on src/Raven.Client/Documents/Indexes/Vector/VectorOptionsFactory.cs:
```diff
@@ -0,0 +1,60 @@
+            }
+        };
+        
+        return new VectorOptionsFactory(newOptions);
```
why are we creating new factory? Don't you want to adjust the options and return this?

maciejaszyk replied:
done

ppekrol commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -1597,6 +1597,67 @@ private void VisitLinqExtensionsMethodCall(MethodCallExpression expression)
+                            textFieldFactory.Invoke(fieldBuilder);
+
+                            if (textFieldValueFactoryObject is not Action<IVectorEmbeddingTextFieldValueFactory> textValueFactory)
+                                throw new Exception();
```
better ex

ppekrol commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -1597,6 +1597,67 @@ private void VisitLinqExtensionsMethodCall(MethodCallExpression expression)
+                            embeddingFieldFactory.Invoke(fieldBuilder);
+
+                            if (embeddingFieldValueFactoryObject is not Action<IVectorEmbeddingFieldValueFactory> embeddingValueFactory)
+                                throw new Exception();
```
better ex

ppekrol commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -1597,6 +1597,67 @@ private void VisitLinqExtensionsMethodCall(MethodCallExpression expression)
+                        }
+                        
+                        default:
+                            throw new Exception();
```
better ex

ppekrol commented on src/Raven.Client/Constants.cs:
```diff
@@ -577,5 +581,23 @@ private SupportedFeatures()
+            internal const string EmbeddingInt8 = EmbeddingPrefix + "i8";
+            internal const string EmbeddingInt1 = EmbeddingPrefix + "i1";
+
+            public const float MinimumSimilarity = 0.8F;
```
DefaultMinimumSimilarity ?

ppekrol commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -1597,6 +1597,67 @@ private void VisitLinqExtensionsMethodCall(MethodCallExpression expression)
+                            fieldFactory.Invoke(fieldBuilder);
+
+                            if (embeddingFieldValueFactoryObject is not Action<IVectorFieldValueFactory> fieldValueFactory)
+                                throw new Exception();
```
better ex

ppekrol commented on src/Raven.Client/Documents/Queries/VectorFieldFactory.cs:
```diff
@@ -0,0 +1,227 @@
+        DestinationQuantizationType = targetEmbeddingQuantization;
+        
+        if (SourceQuantizationType is EmbeddingType.Int8 or EmbeddingType.Binary && DestinationQuantizationType != SourceQuantizationType)
+            throw new InvalidOperationException("Cannot quantize already quantized embeddings.");
```
you need to write those values in the ex, if we will not, user will not know what is going on exactly

ppekrol commented on src/Raven.Client/Documents/Queries/VectorFieldFactory.cs:
```diff
@@ -0,0 +1,227 @@
+            throw new InvalidOperationException("Cannot quantize already quantized embeddings.");
+        
+        if (DestinationQuantizationType == EmbeddingType.Text)
+            throw new InvalidOperationException("Cannot quantize the embedding to Text. This option is only for SourceQuantizationType.");
```
only available

ppekrol commented on src/Raven.Client/Documents/Queries/VectorFieldFactory.cs:
```diff
@@ -0,0 +1,227 @@
+        PortableExceptions.ThrowIf<InvalidOperationException>(_byFieldMethodUsed, $"Cannot use method {nameof(IVectorEmbeddingField.TargetQuantization)} with {nameof(IVectorFieldFactory<T>.WithField)} since quantization is already done by the index.");
+        
+        if (DestinationQuantizationType == EmbeddingType.Text)
+            throw new InvalidDataException("Cannot quantize the embedding to Text. This option is only for SourceQuantizationType.");
```
only available

ppekrol commented on src/Raven.Client/Documents/Queries/VectorFieldFactory.cs:
```diff
@@ -0,0 +1,227 @@
+
+    private static void AssertEmbeddingType<T>()
+    {
+#if NET7_0_OR_GREATER == FALSE
```
#if !NET7_0_OR_GREATER

ppekrol commented on src/Raven.Client/Documents/Session/AbstractDocumentQuery.cs:
```diff
@@ -1513,6 +1514,58 @@ public string GetMemberQueryPath(Expression expression)
             return propertyName;
         }
 
+        internal void VectorSearch(VectorEmbeddingFieldFactory<T> embeddingFieldFactory, VectorFieldValueFactory queriedFactory,
```
queriedFactory? Is that appropriate name?

maciejaszyk replied:
changed

ppekrol commented on src/Raven.Client/Documents/Session/AbstractDocumentQuery.cs:
```diff
@@ -1513,6 +1514,58 @@ public string GetMemberQueryPath(Expression expression)
+                queryParameterName = AddQueryParameter(queriedFactory.Embedding switch
+                {
+                    float[] fa => Convert.ToBase64String(MemoryMarshal.Cast<float, byte>(fa)
+#if NETCOREAPP3_1_OR_GREATER == FALSE
```
use ! for those

ppekrol commented on src/Raven.Client/Documents/Session/AbstractDocumentQuery.cs:
```diff
@@ -1513,6 +1514,58 @@ public string GetMemberQueryPath(Expression expression)
+                    ),
+                    byte[] ba => Convert.ToBase64String(ba),
+                    sbyte[] sb => Convert.ToBase64String(MemoryMarshal.Cast<sbyte, byte>(sb)
+#if NETCOREAPP3_1_OR_GREATER == FALSE
```
use ! for those

ppekrol commented on src/Raven.Server/Documents/Indexes/IndexField.cs:
```diff
@@ -87,7 +94,11 @@ private bool Equals(IndexField other)
+                && SpatialOptions.Equals(Spatial, other.Spatial)
+                && VectorOptions.Equals(Vector, other.Vector)
+                && TermVector == other.TermVector
+                //todo: should we add spatial & vector as well here?
```
I think yes

maciejaszyk replied:
done

ppekrol commented on src/Raven.Server/Documents/Indexes/IndexField.cs:
```diff
@@ -118,6 +129,9 @@ public override int GetHashCode()
                 hashCode = (hashCode * 397) ^ (HasSuggestions ? 233 : 343);
+                hashCode = (hashCode * 397) ^ (Spatial?.GetHashCode() ?? 0);
+                hashCode = (hashCode * 397) ^ (Vector?.GetHashCode() ?? 0);
+                //todo: vector & spatial as well?
```
IMO yes

arekpalinski replied:
@maciejaszyk ? You marked it as resolved but TODO is still here

ppekrol replied:
TODO ^^

ppekrol commented on src/Raven.Server/Documents/Indexes/IndexField.cs:
```diff
@@ -303,6 +341,25 @@ public static string GetSearchAutoIndexFieldName(string name)
+                _ => throw new NotSupportedException($"Unsupported embedding type: ({vectorOptions.SourceEmbeddingType} => {vectorOptions.DestinationEmbeddingType})"),
+            };
+            
+            var inner = innerMethod == string.Empty ? name : $"embedding.{innerMethod}({name})";
```
check if you should not use Constants here that were already created for all those cases

ppekrol commented on src/Raven.Server/Documents/Indexes/IndexStorage.cs:
```diff
@@ -192,8 +192,14 @@ void PersistSearchEngine(Tree configurationTree)
+                    {
+                        // We only support Vectors in Corax, so if an auto-index is using it, let's already set it up as such, regardless
+                        // of what type of default storage engine is configured.
+                        true when _index.Definition.IndexFields.Any(x=> x.Value.Vector != null) => SearchEngineType.Corax,
```
let's talk about this next week

ppekrol replied:
ping

ayende replied:
Did such a talk happen?

ppekrol replied:
We know what to do, but it was not implemented yet.

1. If default auto index engine is corax => no problem
2. if default auto index is lucene then new auto index should be created based on corax, but we need to make adjustments in auto index matcher to avoid accidental sudden switch of queries from one indexing engine to another

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexPersistence.cs:
```diff
@@ -23,7 +23,7 @@ namespace Raven.Server.Documents.Indexes.Persistence.Corax;
 public sealed class CoraxIndexPersistence : IndexPersistenceBase
 {
-    private const bool DisableDictionaryTraining = false; // [DEBUG ONLY]: disable training.
+    private const bool DisableDictionaryTraining = true; // [DEBUG ONLY]: disable training.
```
remember about this

ppekrol reviewed (commented):
Will continue

maciejaszyk committed:
RavenDB-22076
- Renamed EmbeddingType enum for clarity.
- Removed hardcoded constant strings from query parsers.
- Added meaningful exception messages.
- Addressed PR review comments.

maciejaszyk committed:
RavenDB-22076 Added `vector.search` to Studio query assistance.

Lwiel committed:
RavenDB-22076 Fixed quantizer

Lwiel committed:
RavenDB-22076 Post-rebase adjustments

Lwiel committed:
RavenDB-22076 Added StringSegment handling in CoraxQueryBuilder.HandleVector

ppekrol commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -2480,6 +2495,111 @@ private void HandleMoreLikeThis(string methodName, List<QueryExpression> argumen
+                            destinationEmbedding = VectorEmbeddingType.Binary;
+                            break;
+                        default:
+                            throw new InvalidOperationException($"Unknown embedding method type. Supported: [{MethodType.Embedding_Text}, {MethodType.Embedding_Text_I8}, {MethodType.Embedding_Text_I1}, {MethodType.Embedding_F32}, {MethodType.Embedding_F32_I8}, {MethodType.Embedding_F32_I1}, {MethodType.Embedding_I8}, {MethodType.Embedding_I1}] but got `{embeddingType}`.");
```
InvalidQueryException?

maciejaszyk replied:
done

ppekrol commented on src/Raven.Server/Documents/Indexes/VectorSearch/GenerateEmbeddings.cs:
```diff
@@ -0,0 +1,242 @@
+    }
+
+
+#pragma warning disable SKEXP0070 // ignore experimental warning 
```
TODO

ppekrol replied:
Ping

maciejaszyk replied:
removed

ppekrol commented on src/Raven.Server/Documents/Indexes/VectorSearch/GenerateEmbeddings.cs:
```diff
@@ -0,0 +1,242 @@
+        }
+    }
+    
+    private static VectorValue CreateSmartComponentsLocalEmbedding<TEmbedding>(in string text, in int dimensions)
```
is this a correct name?

maciejaszyk replied:
No, changed into `CreateEmbeddingViaSmartComponentsLocalEmbedding`

ppekrol commented on src/Raven.Server/Documents/Indexes/VectorSearch/GenerateEmbeddings.cs:
```diff
@@ -0,0 +1,242 @@
+    private const int F32Size = 1536;
+    
+    [ThreadStatic] // avoid convoys in querying
+    private static ArrayPool<byte> Allocator;
```
isn't that a problem at some point to keep it here without any cleanup?

maciejaszyk replied:
Big buffers should be cleaned via GC (cc @redknightlois ); on the other hand, we're releasing the managed buffer after writing to Corax, so it should reuse the buffer again

ppekrol commented on src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:
```diff
@@ -702,6 +885,14 @@ internal static SpatialField GetOrCreateSpatialField(string name)
             return CurrentIndexingScope.Current.GetOrCreateSpatialField(name);
         }
+        
+        // internal static IndexField GetOrCreateVectorField(string name)
```
TODO

ppekrol replied:
Ping

maciejaszyk replied:
Removed

ppekrol commented on src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:
```diff
@@ -439,13 +626,9 @@ protected IEnumerable<CoraxDynamicItem> CoraxCreateField(CurrentIndexingScope sc
-                Indexing = options?.Indexing,
-            }, allFields, Corax.Constants.IndexWriter.DynamicField);
+
+            var field = IndexField.Create(name, new IndexFieldOptions { Storage = options?.Storage, TermVector = options?.TermVector, Indexing = options?.Indexing, },
```
what is the change here?

ppekrol replied:
Ping

maciejaszyk replied:
Formatting, restored old one.

ppekrol commented on src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:
```diff
@@ -330,6 +338,185 @@ protected dynamic TryConvert<T>(object value)
+
+            VectorValue GenerateEmbeddingsFromDynamicArray<TEnumerable>(TEnumerable dynamicArray) where TEnumerable : IEnumerable
+            {
+                //Todo: we could use some kind of growable buffer instead list to avoid allocations
```
TODO

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs:
```diff
@@ -30,6 +36,7 @@
 using RavenConstants = Raven.Client.Constants;
 using IndexSearcher = Corax.Querying.IndexSearcher;
 using CoraxConstants = Corax.Constants;
+using Memory = NetTopologySuite.Utilities.Memory;
```
?

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexedEntriesReader.cs:
```diff
@@ -98,7 +100,36 @@ public DynamicJsonValue GetDocument(ref EntryTermsReader entryReader)
+                        var blit = new BlittableJsonReaderObject(span.Address, span.Length, _ctx);
+                        SetValue(fieldName, blit);
+                    }
+                    catch
```
what case this handles? why are we handling this via catch?

maciejaszyk committed:
RavenDB-22076
- Avoided managed allocation
- Moved VectorValue into Corax project
- Partially redesigned StaticIndexBase method for VectorSearch

maciejaszyk committed:
RavenDB-22076 Minor fixes

Lwiel committed:
RavenDB-22076 Added XML documentation of VectorSearch

maciejaszyk committed:
RavenDB-22076 Fixed a NullReferenceException occurring when querying Corax's index with VectorSearch/PhraseQuery when the index contains no documents.

maciejaszyk committed:
RavenDB-22076 WIP: Initial tests for similarity (via query).

ayende committed:
RavenDB-22076
- Properly implementing HNSW + optimizations
- Added Exact Nearest Neighbor search
- Add a debug output that can visualize a search through the HNSW graph
- Adding benchmark that can test that we write & read properly for vector search

Lwiel committed:
RavenDB-22076 Fixed quantization tests

ayende committed:
RavenDB-22076
- Fixing benchmark to verify recall on real world dataset
- Refactoring API to use a named HSNW instead of anonymous one
- We now have a single global container for vectors (indexed by blake2d hash)
- Each HNSW (field) has a list of nodes by vector and nodes by vector id (the global value)
- Temporarily disable LoadNodeIndexes to simpler impl
- Fixed NearestEdges (need to inverse PQ to work properly)
- Added a way to compute distances as well during lookup

ayende committed:
RavenDB-22076
- Properly handle batch load of nodes during reads

ayende committed:
RavenDB-22076
- Reduce managed allocations
- Always fill the distances

ayende committed:
RavenDB-22076
- Avoid holding a ref reference to a potentially re-alloced array, instead pass the index in the array

ayende committed:
RavenDB-22076
- Do no return results of nodes with empty posting lists

ayende committed:
RavenDB-22076
- Implementing vector storage to reduce fragmentation costs

ayende committed:
RavenDB-22076
- Refactoring how we are implementing vector batches
- Removing debug code

ayende committed:
RavenDB-22076
- Implementing removal from the HNSW

ayende committed:
RavenDB-22076
 - Fixing insertion of the second large vector id

ayende committed:
RavenDB-22076
- Handle large posting lists in HNSW

Lwiel committed:
RavenDB-22076 Added vector dimensions check for indexing

maciejaszyk commented on src/Raven.Server/Documents/Indexes/IndexFieldsPersistence.cs:
```diff
@@ -29,6 +33,28 @@ internal void Initialize()
+                    {
+                        // In case of VectorEmbeddingType.Single the embedding length is multiplied by 4 (for every float we have 4 bytes), so to match this behavior we 
+                        // have to multiply the length here by 4
+                        VectorEmbeddingType.Single => fieldDimensions.Value * 4,
```
``` * sizeof(float)```

maciejaszyk commented on src/Raven.Server/Documents/Indexes/IndexFieldsPersistence.cs:
```diff
@@ -50,21 +76,64 @@ internal bool HasTimeValues(string fieldName)
+                var (originalStoredDimensions, originalDimensions) = destinationEmbeddingType switch
+                {
+                    // 4 = sizeof(float) / sizeof(byte)
+                    VectorEmbeddingType.Single => (storedDimensions / 4, dimensions / 4),
```
Shouldn't it be multiplied? For a vector with 1000 dimensions, we will be using `1000 * sizeof(float)` bytes.

Lwiel committed:
RavenDB-22076 Added comments to embedding dimensions check

maciejaszyk committed:
RavenDB-22076
- Introduced masks as constants for readability.
- Bugfix: During a node update without changing the node's disk location, write it to the container buffer.

maciejaszyk committed:
RavenDB-22076
- Added support for removal based on vector hash instead of raw vector.
- Fixed a bug with skipping removed elements.
- Added support for additional similarity distance.

Lwiel committed:
RavenDB-22076 Added NumberOfCandidates and NumberOfEdges options to client

maciejaszyk committed:
RavenDB-22076
- Initial work to adapt HNSW in Corax
- Introduced VectorOptions configuration in Corax
- HNSW: Moved commit logic to a dedicated method

maciejaszyk committed:
RavenDB-22076 Changed `NearestSearch` from beign a `ref struct` to a `struct` for querying purposes.

Lwiel committed:
RavenDB-22076 Adjusted Server code to changes in HNSW

maciejaszyk committed:
RavenDB-22076 Extended `GrowableBuffer` to be generic storage.

maciejaszyk committed:
RavenDB-22076
- Removed old code that used stored vector in the index entry.
- Introduced a new Corax query primitive to handle HNSW search.
- Introduced a dedicated EntryIdToInternalEntryId to fulfill the requirement `entryId & 0b11 == 0`.
- Fixed old vector tests.
- HNSW can use different distance methods for different vector representations.
- Handle the case when there is no vector in the graph.

Lwiel committed:
RavenDB-22076 Added magnitude storage for I8 quantized vectors

maciejaszyk committed:
RavenDB-22076
- Passed allocator for embedding generator
- Dispose of embedding in querying
- Conditional pass of random to create reproducible test for HNSW
- Test adjustments and fixes

maciejaszyk committed:
RavenDB-22076 Testfix: clone vector hash to managed memory to ensure persisting correct hash for vector removal.

maciejaszyk committed:
RavenDB-22076 Skip first byte of the container storage for count (for first vector).

maciejaszyk committed:
RavenDB-22076
- Throw when `minimumSimilarity` is not in the (0, 1] range.
- Low-level test with real vector values
- Parameter adjustments in tests

maciejaszyk committed:
RavenDB-22076
- Remove count (offset + 1) when writing the raw vector since the count is encoded in the vector id.
- When writing the score, we've to transform the distance  to score (bigger is closer)
- Test adjustments

maciejaszyk committed:
RavenDB-22076 Test for nearest search

maciejaszyk committed:
RavenDB-22076 Register postinglist container into PostingList register for storage calculation purposes.

maciejaszyk committed:
RavenDB-22076
- Persist the `FieldRootPage` of the Vectors field for `EntryReader` purposes
- Move `Hnsw.Options` into a dedicated file
- Move the constants of `Hnsw` into a dedicated file
- Fix the storage report when there is a vector field
- Remove the unused enumerable of entry readers
- Querying: correct the parsing of `minimumMatch`
- `ValueReader`: get `UnmanagedSpan<T>` from the underlying value

maciejaszyk committed:
RavenDB-22076
- Removed unused code
- Turn on dictionary training
- Test fix

maciejaszyk committed:
RavenDB-22076
- Support for ExactSeach from LINQ/DocumentQuery
- Defines color for VectorValues in IndexingPerformance view.

Lwiel committed:
RavenDB-22076 Documentation of LINQ extensions

Lwiel committed:
RavenDB-22076 Fixed auto index naming on exact search

arekpalinski commented on src/Corax/Utils/EntryTermsReader.cs:
```diff
@@ -409,6 +425,7 @@ public string Debug(Dictionary<long, string> fields = null)
                 continue;
             }
 
+            // TODO: Handle vector here
```
TODO

maciejaszyk replied:
done

ayende replied:
Still here

arekpalinski commented on src/Raven.Client/Documents/Queries/VectorFieldFactory.cs:
```diff
@@ -0,0 +1,267 @@
+using Sparrow;
+using Sparrow.Binary;
+
+namespace Raven.Client.Documents.Queries;
```
As I see we use `Raven.Client.Documents.Indexes.Vector` namespace for indexing related classes, but querying related item are put directly in `Raven.Client.Documents.Queries`. Shouldn't we use `Raven.Client.Documents.Queries.Vector`?

maciejaszyk replied:
It should. All classes related to vectorsearch have been moved.

arekpalinski commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -5215,6 +5217,30 @@ private SpatialField GetOrAddSpatialField(string name)
+                
+                return field switch
+                {
+                    //todo: revisit it later
```
TODO

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxDocumentConverterBase.cs:
```diff
@@ -408,10 +421,13 @@ protected void InsertRegularField<TBuilder>(IndexField field, object value, Json
-                //we want to unpack item here.
                 var old = builder.ResetList(); // For lists of CreatedField(), we ignoring the list
-                InsertRegularField(cdi!.Field, cdi.Value, indexContext, builder, sourceDocument, out shouldSkip);
+                if (value is CoraxDynamicItem standardCdi)
```
We didn't have this check before. Why do we have it now? When this cannot be true?

ppekrol replied:
Ping

maciejaszyk replied:
Restored the previous code, this was changed only for debugging purposes.

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs:
```diff
@@ -612,6 +625,108 @@ IQueryMatch HandleStringUnaryMatch(Parameters queryEnvironment)
+        {
+            transformedEmbedding = default;
+            // var embeddings = GenerateEmbeddings.FromArray(vectorOptions, Convert.FromBase64String(stringSegment.ToString()));
+            // transformedEmbedding = embeddings.EmbeddingAsBytes;
```
Commented code

arekpalinski commented on src/Raven.Server/Documents/Queries/LuceneQueryBuilder.cs:
```diff
@@ -411,6 +411,9 @@ private static Lucene.Net.Search.Query ToLuceneQuery(TransactionOperationContext
                         return new MatchAllDocsQuery();
+                    case MethodType.Vector_Search:
+                        throw new InvalidQueryException("Vector Search is not available for the Lucene indexing engine, only in the Corax indexing engine.", query.QueryText, parameters);
+                        break;
```
Please check the warning here

arekpalinski commented on src/Voron/Data/Graphs/Hnsw.Debug.cs:
```diff
@@ -0,0 +1,161 @@
+                var item = Container.Get(llt, postingListId & ~0b11);
+                return VariableSizeEncoding.Read<int>(item.Address, out _);
+            case 0b11:
+                throw new NotImplementedException();
```
Let's provide messages to NotImplementedException() we have in this file

ayende replied:
That should not be not implemented. That should be reading the count from the actual posting list here.

maciejaszyk replied:
Implemented.

arekpalinski commented on src/Voron/Data/Graphs/Hnsw.cs:
```diff
@@ -0,0 +1,1315 @@
+            if (vector.IsEmpty)
+            {
+                ref var from = ref GetNodeByIndex(fromIdx);
+                vector = from.GetVector(in this);
```
Please check the warning here

arekpalinski commented on src/Voron/Data/Graphs/Hnsw.cs:
```diff
@@ -0,0 +1,1315 @@
+        
+        public void Dispose()
+        {
+            //todo: we may wants to release the vector hash cache
```
TODO

maciejaszyk replied:
@redknightlois 
Can I safely release the memory of the ByteString keys from dictionary?
eg.
```csharp
            foreach (var key in _vectorHashCache.Keys)
            {
                var k = key;
                _searchState.Llt.Allocator.Release(ref k);
            }
            
            _vectorHashCache.Clear();
```

redknightlois replied:
You can release the memory of the `ByteString` keys ***if and only if*** you adhere to these principles:

**Lifecycle Ownership:** The moment you put something into the dictionary, the dictionary becomes its owner. This means you relinquish all other references. If you kept a reference somewhere else, you’re setting yourself up for undefined behavior—so don’t.

**Transaction Context:** The release must happen in the same transaction where the keys were allocated. Cross-transaction operations with memory management? That’s a recipe for disaster.

**Iterator Behavior:** Iterators need to play nice. If the iterator is going to touch or dereference keys they must not have been disposed, and the one that iterates must not keep those pointer.

arekpalinski commented on src/Voron/Data/Graphs/Hnsw.cs:
```diff
@@ -0,0 +1,1315 @@
+            return index;
+        }
+    }
+}
```
It's pretty big file. Can we move nested classes / structs to separate files?

maciejaszyk replied:
Agree, but let's do it at the end of the work since it's easier to track the history of changes during development.

arekpalinski commented on test/FastTests/Corax/Vectors/IndexVectorExactViaStaticIndexesTests.cs:
```diff
@@ -0,0 +1,301 @@
+                select new { Id = doc.Id, Vector = CreateVectorSearch(doc.Text) };
+
+
+            VectorIndexes.Add(x => x.Vector,
```
Why did we choose `VectorIndexes` name? Maybe just `Vectors` or `SearchVectors` (to distinguish from `TermsVectors` that we already have)?

maciejaszyk replied:
I actually copied the naming from spatial fields for API naming consistency.

maciejaszyk committed:
RavenDB-22076
- Adjusted the studio DTO model of VectorOptions
- Prevented Hnsw.Create from beign called more than once per batch, per field
- Fixed the vectorized code for transforming internal HNSW entryId to the input entry
- StorageReportGenerator: We must check if our element fits the Hnsw.Options structure before checking the RootType
- Removed the container from `StoredValues` after removing the Vector from the graph.

maciejaszyk commented on src/Raven.Client/Documents/Indexes/Vector/AutoVectorOptions.cs:
```diff
@@ -0,0 +1,41 @@
+    {
+        
+    }
+    public AutoVectorOptions(AutoVectorOptions options)
+    {
+        Dimensions = options.Dimensions;
+        SourceEmbeddingType = options.SourceEmbeddingType;
+        DestinationEmbeddingType = options.DestinationEmbeddingType;
+        SourceFieldName = options.SourceFieldName;
+    }
```
We have global and constant values for `NumberOfEdges`/`NumberOfCandidates`. Do we want to introduce configuration options for them (not granularly, just for auto indexes in the database)?

cc @ayende

ayende replied:
Yes, that would be good to have.
Would apply to auto indexes only? Or the default (unless overwritten) in general?

maciejaszyk commented on src/Raven.Client/Documents/Session/AbstractDocumentQuery.cs:
```diff
@@ -1513,6 +1514,57 @@ public string GetMemberQueryPath(Expression expression)
+            else if (embeddingValueFactory.Embedding != null)
+            {
+                // for well-known types we can convert the array into Base64
+                queryParameterName = AddQueryParameter(embeddingValueFactory.Embedding switch
+                {
+                    float[] fa => Convert.ToBase64String(MemoryMarshal.Cast<float, byte>(fa)
+#if !NETCOREAPP3_1_OR_GREATER
+                        .ToArray() // For newer frameworks we use Span overload in Convert.
+#endif
+                    ),
+                    byte[] ba => Convert.ToBase64String(ba),
+                    sbyte[] sb => Convert.ToBase64String(MemoryMarshal.Cast<sbyte, byte>(sb)
+#if !NETCOREAPP3_1_OR_GREATER
+                        .ToArray()
+#endif
+                    ),
+                    _  => embeddingValueFactory.Embedding,
```
TODO benchmark

Currently, parsing a float array from a blittable array is expensive, so we could just send the base64-encoded array. However, once https://github.com/ravendb/ravendb/pull/19603 is completed, we can send the floats in vector array form.

maciejaszyk commented on src/Raven.Client/Documents/Session/IDocumentQueryBase.cs:
```diff
@@ -440,6 +440,36 @@ public interface IFilterDocumentQueryBase<T, TSelf> where TSelf : IDocumentQuery
+        TSelf VectorSearch(Func<IVectorFieldFactory<T>, IVectorEmbeddingField> embeddingFieldFactory, Action<IVectorEmbeddingFieldValueFactory> embeddingValueFactory, float minimumSimilarity = Constants.VectorSearch.DefaultMinimumSimilarity, int numberOfCandidates = Constants.VectorSearch.DefaultNumberOfCandidatesForQuerying, bool isExact = Constants.VectorSearch.DefaultIsExact);
+        
+        /// <summary>
+        /// Performs vector search on existing vector field.
```
@Lwiel 
maybe we should add some info about index, because it may be misleading with document's field.

Lwiel replied:
Done

Lwiel committed:
RavenDB-22076 Changed debug assertion

maciejaszyk committed:
RavenDB-22076
- Allow index null in vector field
- Extended debug methods for vectors
- Moved VectorSearch related class into `Raven.Client.Documents.Queries.Vector`
- Removed unused usings
- PR comments
- Added `NumberOfEdges` & `NumberOfCandidates` into JsonSerialization methods for `VectorOptions`
- Added StressTests for guarding maximal number of vector in self managed container for Vector storage.
- Changed SearchState in HNSW into class

ayende commented on src/Corax/Indexing/IndexWriter.cs:
```diff
@@ -880,6 +931,13 @@ public void Commit<TStatsScope>(TStatsScope stats)
 
+                if (indexedField.VectorIndexWriter != null)
+                {
+                    using var __ = staticFieldScope.For(CommitOperation.VectorValues);
+                    RegisterVectorRootPage(indexedField.FieldRootPage);
+                    indexedField.VectorIndexWriter.Commit();
```
What is that about?

maciejaszyk replied:
Previously, we did the writing in the `Dispose` method, I've moved the body into `Commit` for semantic reason.

ayende commented on src/Corax/Indexing/IndexWriter.cs:
```diff
@@ -459,17 +465,27 @@ private void ProcessDeletes()
+                {
+                    bool exists = fieldsByRootPage.TryGetValue(reader.FieldRootPage, out var field);
+                    PortableExceptions.ThrowIfNot<InvalidOperationException>(exists, "Tried to remove vector but couldn't find the associated indexed field.");
+                    var vectorWriter = field!.GetVectorWriter(_transaction.LowLevelTransaction);
```
`GetVectorWriter` is really misleading name here

maciejaszyk replied:
Maybe better name is `VectorIndexer`?

ayende commented on src/Corax/Querying/Matches/VectorSearchMatch.cs:
```diff
@@ -0,0 +1,204 @@
+
+    public SkipSortingResult AttemptToSkipSorting()
+    {
+        return SkipSortingResult.ResultsNativelySorted;
```
They aren't, though, are they? 
The results you get are from the various posting lists, and aren't sorted

maciejaszyk replied:
You're right, changed into `SkipSortingResult.SortingIsRequired`

ayende commented on src/Corax/Querying/Matches/VectorSearchMatch.cs:
```diff
@@ -0,0 +1,204 @@
+        return amountToCopy;
+    }
+
+    private int FillDiscardSimilarity(Span<long> matches)
```
I'm not following what this is for

maciejaszyk replied:
When you have a query like this:
`from index MyIndex where [WHERE CLAUSE] or vector.search(MyVector, $p0, 0.7)`
you don't actually care about the similarity values. Therefore, if the query doesn't include `order by score`, we're not persisting the distances, which saves memory.

ayende commented on src/Corax/Querying/Matches/VectorSearchMatch.cs:
```diff
@@ -0,0 +1,204 @@
+            return;
+        }
+        
+        _maximumDistance = _nearestSearch.MinimumSimilarityToMaximumDistance(minimumMatch);
```
I don't understand this. Why go to this trouble and not simple pass the `minimumMatch` to the `Hnsw` ?

maciejaszyk replied:
done

ayende commented on src/Corax/Querying/Matches/VectorSearchMatch.cs:
```diff
@@ -0,0 +1,204 @@
+        
+        ref var matchesRef = ref MemoryMarshal.GetReference(matches);
+        ref var scoresRef = ref MemoryMarshal.GetReference(scores);
+        ref var distanceRef = ref MemoryMarshal.GetReference(_distances.Results);
```
The idea here is that we need to score relatively few numbers, right?
So the cost of doing all those binary searches is small?

maciejaszyk replied:
That's the idea. We have to find the positions anyway, and since incoming matches can have many documents coming from different primitives, I would say it's okay to perform a binary search. We do something similar for boosting in other primitives.

ayende replied:
What about when we are ordering by this score?
Won't that force us to go over the entire thing?

maciejaszyk replied:
Eventually, yes, but we do not have any information about whether we have other primitives or not, so it is required.

ayende commented on src/Corax/Utils/VectorValue.cs:
```diff
@@ -0,0 +1,73 @@
+{
+    private readonly bool _isNative;
+    private readonly ArrayPool<byte> _bufferPool;
+    private byte[] _managedBuffer;
+    private Memory<byte> _managedMemory;
+    private ByteString _nativeMemory;
```
Why all of those?

maciejaszyk replied:
When we generate the embedding via the library, it is inside managed memory. However, when we're reading from a document, we use native memory as a holder. The `VectorValue` is a struct that abstracts the vector value for indexing.

ayende commented on src/Raven.Client/Documents/Indexes/Vector/VectorOptions.cs:
```diff
@@ -0,0 +1,109 @@
+            hashCode = (hashCode * 397) ^ Dimensions.GetHashCode();
+            
+            return hashCode;
+        }
```
Where are we storing this in a hash table?

maciejaszyk replied:
I do not recall the place, but all other `[...]Options` had it implemented, so I follow the convention.

ayende commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -1597,6 +1597,76 @@ private void VisitLinqExtensionsMethodCall(MethodCallExpression expression)
+                    
+                    LinqPathProvider.GetValueFromExpressionWithoutConversion(expression.Arguments[3], out var minimumSimilarityObject);
+
+                    if (minimumSimilarityObject is not float minimumSimilarity || minimumSimilarity is < -1.0f or > 1.0f)
```
Is `minimumSimilarityObject` required?
What error is raised if it isn't provided?

maciejaszyk replied:
It's not required. Currently, it's nullable field (server will replace null with the default value from configuration).
A `InvalidDataException` will be thrown when `minimumSimilarity` is out of range.

`NumberOfCandidates` is handled in similar manner.

ayende commented on src/Raven.Client/Documents/Linq/RavenQueryProviderProcessor.cs:
```diff
@@ -1597,6 +1597,76 @@ private void VisitLinqExtensionsMethodCall(MethodCallExpression expression)
                 case nameof(LinqExtensions.Skip):
                     VisitQueryableMethodCall(expression);
                     break;
+                case nameof(LinqExtensions.VectorSearch):
```
Add a comment here showing the expected calls forms

maciejaszyk replied:
Added

ayende commented on src/Raven.Server/Documents/Indexes/IndexFieldsPersistence.cs:
```diff
@@ -29,6 +33,29 @@ internal void Initialize()
+                        // In case of VectorEmbeddingType.Single the embedding length is multiplied by 4 (for every float we have 4 bytes), so to match this behavior we 
+                        // have to multiply the length here
+                        VectorEmbeddingType.Single => fieldDimensions.Value * sizeof(float),
+                        VectorEmbeddingType.Int8 => fieldDimensions.Value,
```
Don't you need + 4 here, for the scale factor? Or is it the raw size already?

Lwiel replied:
Yes, we need it since we store the magnitude now
Added

ayende commented on src/Raven.Server/Documents/Indexes/IndexStorage.cs:
```diff
@@ -192,8 +192,14 @@ void PersistSearchEngine(Tree configurationTree)
+                        true when _index.Definition.IndexFields.Any(x=> x.Value.Vector != null) => SearchEngineType.Corax,
+                        true => _index.Configuration.AutoIndexingEngineType,
+                        false => _index.Configuration.StaticIndexingEngineType
+                    };
```
where do we error if the user created a static index usign Lucene with vectors?

maciejaszyk replied:
Currently, we throw in document converter (no support of `Vector` type).

ayende commented on src/Raven.Server/Documents/Indexes/IndexStorage.cs:
```diff
@@ -1197,6 +1203,40 @@ internal void WriteIndexTimeFields(RavenTransaction tx, HashSet<string> timeFiel
 
+        internal Dictionary<string, int> ReadVectorDimensions()
+        {
+            Dictionary<string, int> container = new();
```
That is `(string FieldName, int VectorSizeInBytes)` ?

Lwiel replied:
Yes, I added the comment

ayende commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs:
```diff
@@ -612,6 +625,108 @@ IQueryMatch HandleStringUnaryMatch(Parameters queryEnvironment)
+            };
+        }
+
+        var numberOfCandidates = 12;
```
Should be constant - or maybe configuration?

maciejaszyk replied:
Default value moved into configuration

ayende commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxQueryBuilder.cs:
```diff
@@ -612,6 +625,108 @@ IQueryMatch HandleStringUnaryMatch(Parameters queryEnvironment)
+            transformedEmbedding = GenerateEmbeddings.FromArray(vectorOptions, memScope, mem, bytesUsed);
+        }
+        
+        var minimumMatch = RavenConstants.VectorSearch.DefaultMinimumSimilarity;
```
Maybe configuration?

maciejaszyk replied:
Added configuration for:
- Default MinimumSimilarity
- Default NumberOfCandidates for Indexing & Querying
- Default NumberOfEdges for Indexing.

ayende commented on src/Raven.Server/Documents/Indexes/VectorSearch/GenerateEmbeddings.cs:
```diff
@@ -0,0 +1,194 @@
+                }
+                else
+                {
+                    VectorQuantizer.TryToInt8(source, dest, out usedBytes);
```
Can the `Try` method fail?

maciejaszyk replied:
Actually, yes, when the `dest` buffer does not have enough space.

ayende commented on src/Raven.Server/Documents/Indexes/VectorSearch/GenerateEmbeddings.cs:
```diff
@@ -0,0 +1,194 @@
+        var currentAllocator = Allocator ??= ArrayPool<byte>.Create();
+        var buffer = currentAllocator.Rent(dimensions);
+        var embedding = new Memory<byte>(buffer, 0, dimensions);
+        Embedder.Embed<TEmbedding>(text, embedding);
```
Where do we store what model we use?
What if we want to use a different one?

maciejaszyk replied:
https://github.com/dotnet-smartcomponents/smartcomponents/blob/main/docs/local-embeddings.md#customizing-the-underlying-embeddings-model

The model is defined during compilation time.

ayende replied:
Sure, and now consider that models are something that are often changed.
In 2 years, we want to use a better default model. How do we do that?

ppekrol replied:
TODO: needs discussion

maciejaszyk replied:
https://github.com/dotnet/smartcomponents/blob/7de0258fee1283f63f5a4064a0391a5651d2cae4/src/SmartComponents.LocalEmbeddings/LocalEmbedder.cs#L32-L50

It seems the library allows defining the model from the path.

maciejaszyk committed:
RavenDB-22076 Fixed compilation errors and adjust methods after rebase.

maciejaszyk committed:
RavenDB-22076:
- Remove duplicates from `Fill` calls in `VectorSearchMatch`
- Move `MinimumSimilarity` into `HNSW` search

maciejaszyk committed:
RavenDB-22076
- Moved VectorSearch querying constants into configuration.
- Default values of `minimumSimilarity` and `numberOfCandidates` is `null`.
- Fixed hash code calculation of vector options.
- Default values for the number of candidates and the number of edges are read from configuration when not explicitly configured.
- Added new configuration keys into the studio.
- Added null handling in the VectorOptions JSON writer.

maciejaszyk committed:
RavenDB-22076
- Throw better exception when properties of vector search are not acceptable (incorrect type / invalid values)
- Handle StringSegments during query builder phase.

Lwiel committed:
RavenDB-22076 Fixed dimensions check and XML comment

Lwiel committed:
RavenDB-22076 Added more tests and comments

maciejaszyk committed:
RavenDB-22076 When searching for the nearest edges, we must ensure that the destination point is not added to the `nearestEdges` queue to avoid a loop in the path

maciejaszyk committed:
RavenDB-22076 `GrowableBuffer - Constant`: `initSize` defines the number of elements of `typeof(T)` without considering memory usage

ppekrol commented on RavenDB.sln:
```diff
@@ -749,6 +751,24 @@ Global
+		{A24030B6-A51C-47B5-8291-DD1DD4B2DA4E}.Release|x64.Build.0 = Release|Any CPU
+		{A24030B6-A51C-47B5-8291-DD1DD4B2DA4E}.Release|x86.ActiveCfg = Release|Any CPU
+		{A24030B6-A51C-47B5-8291-DD1DD4B2DA4E}.Release|x86.Build.0 = Release|Any CPU
+		{A24030B6-A51C-47B5-8291-DD1DD4B2DA4E}.Validate|Any CPU.ActiveCfg = Debug|Any CPU
```
Debug|Any CPU => Release|Any CPU

ppekrol commented on Directory.Packages.props:
```diff
@@ -97,6 +98,7 @@
     <PackageVersion Include="System.Linq.Async" Version="6.0.1" />
     <PackageVersion Include="System.Memory" Version="4.6.0" />
     <PackageVersion Include="System.Net.Http" Version="4.3.4" />
+    <PackageVersion Include="System.Numerics.Tensors" Version="9.0.0-rc.2.24473.5" />
```
after rebase you can switch to 9.0.0

maciejaszyk replied:
done

ppekrol commented on bench/Vector.Benchmark/Vector.Benchmark.csproj:
```diff
@@ -0,0 +1,19 @@
+
+    <PropertyGroup>
+        <OutputType>Exe</OutputType>
+        <TargetFramework>net8.0</TargetFramework>
```
after rebase switch that to net9.0

maciejaszyk replied:
done

ppekrol commented on src/Raven.Client/Documents/Queries/Vector/VectorQuantizer.cs:
```diff
@@ -0,0 +1,85 @@
+
+namespace Raven.Client.Documents.Queries.Vector;
+
+public class VectorQuantizer
```
static?

maciejaszyk replied:
changed

ppekrol commented on src/Raven.Server/Documents/Indexes/IndexField.cs:
```diff
@@ -87,7 +94,10 @@ private bool Equals(IndexField other)
+                && SpatialOptions.Equals(Spatial, other.Spatial)
+                && VectorOptions.Equals(Vector, other.Vector)
+                && TermVector == other.TermVector
+                ;
```
:)

maciejaszyk committed:
RavenDB-22076
- Finished TODOs
- Reverted unnecessary changes made during development
- Renamed `VectorWriter` to `VectorIndexer`
- Unified `CreateVector` methods
- Changed project validation from `Debug|Any CPU` to `Release|Any CPU`
- `VectorSearchMatch`: results must be sorted.

maciejaszyk committed:
RavenDB-22076
- Updated `Vector.Benchmark` to `.net9`
- Updated `System.Numerics.Tensor` to `9.0.0`

maciejaszyk committed:
RavenDB-22076 AutoVectorOptions: copy `NumberOfEdges` and `NumberOfCandidatesForIndexing`

maciejaszyk committed:
RavenDB-22076 Fixed conventions in the code.
